### PR TITLE
[RELAY] Fixes to MergeCompilerRegions

### PR DESCRIPTION
**Pull request by mbaret**
There were a few outstanding issues with the previous PR to implement the MergeCompilerRegions pass. This PR addresses those issues and fixes the AnnotateTarget test.

### CODE REVIEW

**Comment by mbaret** [comment]
cc @zhiics @trevor-m @comaniac 
Can you see if you still observe failures with this PR?

**Comment by trevor-m** [comment]
Thanks @mbaret ! This fixes the issues with resnet and other networks.

I did notice however that `AnnotateRestDefault` labels the ops meant to be run in TVM as "relay.ext.default" which results in a failure during codegen since there is obviously no "default" codegen: `TVMError: Check failed: pf: Failed to find the codegen tool for relay.ext.default`. Am I supposed to do something extra to remove the default label after partitioning?

**Comment by zhiics** [comment]
For "default", I commented in the other PR: https://github.com/apache/incubator-tvm/pull/5028

File: src/relay/transforms/annotate_target.cc [hunk 1]
```diff
@@ -52,6 +55,13 @@ class AnnotateTargetWrapper : public ExprMutator {
         return fannotate[op](call->attrs, call->args);
       }
     }
+    if (expr->IsInstance<TupleGetItemNode>()) {
```
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc]: I have a concern about this. It seems like you have to list all possible nodes here and it's easy to miss something. For now I would suggest adding an exception at least to indicate that we need to add more here.

File: src/relay/transforms/merge_compiler_regions.cc [hunk 2]
```diff
@@ -239,10 +258,14 @@ class RegionMerger : public ExprVisitor {
   void VisitExpr_(const CallNode* call) final {
     if (call->op == compiler_end_op) {
       auto region = regions_->GetRegion(GetRef<Call>(call));
+      auto node = (*region->GetOutputs().begin()).as<CallNode>();
+      std::string name = "";
```
Comment (comaniac) [inline src/relay/transforms/merge_compiler_regions.cc]: What is this `name` for?

File: src/relay/transforms/annotate_target.cc (hunk 1 above)
Comment (trevor-m) [reply src/relay/transforms/annotate_target.cc]: I have a PR ready to add support for TupleNode.

File: src/relay/transforms/merge_compiler_regions.cc [hunk 3]
```diff
@@ -86,126 +86,145 @@ class AnnotateRestDefault : public ExprMutator {
...
+    return expr;
   }
 
-  Expr AddCompilerEnd_(const Expr& expr) {
-    const auto* end_op =
-      runtime::Registry::Get("relay.op.annotation._make.compiler_end");
-    CHECK(end_op);
-    Expr end = (*end_op)(expr, target_);
-    return end;
+  /*! \brief This function adds compiler begins to nodes that
```
Comment (zhiics) [inline src/relay/transforms/merge_compiler_regions.cc:105]: Alignment

File: src/relay/transforms/merge_compiler_regions.cc [hunk 4]
```diff
@@ -86,126 +86,145 @@ class AnnotateRestDefault : public ExprMutator {
...
-    // AND they should not be arguments of the original function
-    if (!region.defined() &&
-       std::find(func_->params.begin(),
-                 func_->params.end(), visited_expr)
-                   == func_->params.end()) {
-      return AddCompilerEnd_(visited_expr);
+  Expr InsertEnd(const Expr& expr) {
+    if (annotated_nodes_.find(expr) == annotated_nodes_.end() &&
+        !expr->IsInstance<VarNode>() && !expr->IsInstance<ConstantNode>()) {
+      const auto *end_op =
```
Comment (zhiics) [inline src/relay/transforms/merge_compiler_regions.cc:96]: const auto*

File: src/relay/transforms/merge_compiler_regions.cc [hunk 5]
```diff
@@ -86,126 +86,145 @@ class AnnotateRestDefault : public ExprMutator {
...
-    CHECK(end_op);
-    Expr end = (*end_op)(expr, target_);
-    return end;
+  /*! \brief This function adds compiler begins to nodes that
+ * don't belong to a region already (default).
+ * \param expr The expression to add a compiler begin to.
+ * \return expr The expression with or without a compiler begin added.
+ */
+  Expr InsertBegin(const Expr& expr) {
+    const auto *begin_op =
```
Comment (zhiics) [inline src/relay/transforms/merge_compiler_regions.cc:111]: const auto*

**Comment by trevor-m** [comment]
HI @mbaret, I prepared this fix to include TupleNode in annotation when surrounded by supported nodes, in a similar way to what you have done for TupleGetItemNode.

Here is the code, you are welcome to include it in this PR if you would like.
https://github.com/trevor-m/tvm/commit/2b7c3b1d040abf1f88fca3857c65f13c4f012b2e

File: src/relay/transforms/merge_compiler_regions.cc (hunk 2 above)
Comment (mbaret) [reply src/relay/transforms/merge_compiler_regions.cc]: Debugging debris.. fixed.

File: src/relay/transforms/annotate_target.cc (hunk 1 above)
Comment (mbaret) [reply src/relay/transforms/annotate_target.cc:58]: I'm not sure yet what we plan on doing with Let nodes (or Ref* nodes). I think just saying they're not supported is sufficient for now though without also throwing an error.

**Comment by mbaret** [comment]
@trevor-m @zhiics Default behaviour should now be fixed.

File: tests/python/relay/test_pass_merge_compiler_regions.py [hunk 6]
```diff
@@ -66,13 +66,13 @@ def expected():
         O_2 = relay.nn.relu(O_1)
         ce_3 = compiler_end(O_2, "test")
 
-        cb_x = compiler_begin(ce_2, "default")
-        X = relay.tanh(cb_x)
-        ce_x1 = compiler_end(X, "default")
-        ce_x2 = compiler_end(X, "default")
+        # cb_x = compiler_begin(ce_2, "default")
```
Comment (zhiics) [inline tests/python/relay/test_pass_merge_compiler_regions.py:69]: remove the commented lines

**Comment by zhiics** [comment]
Can you add a unit test for end-to-end test? It should include anntation-mege-partitioning and we can invoke it using runtime. I just want to make sure everything is correct.

**Comment by mbaret** [comment]
> Can you add a unit test for end-to-end test? It should include anntation-mege-partitioning and we can invoke it using runtime. I just want to make sure everything is correct.

Yup, I'm looking at adding this right now :) My question is what codegen can I use?

**Comment by zhiics** [comment]
You can have a mock example that uses gcc. I think we have examples for it already.

**Comment by mbaret** [comment]
Do you mean ccompiler? If so, I don't think that supports codegen on a region of multiple ops.

**Comment by zhiics** [comment]
hmm, you are right. Let's have examples with annotation-merge-partitioning with structural equal check first.

**Comment by trevor-m** [comment]
Hi @mbaret, thanks again for fixing these issues!

I noticed the call to MergeCompilerRegions hangs indefinitely for Densenet121 when all nodes are annotated. Here is a script to reproduce the issue. There might a some infinite loop in the pass code or the runtime is scaling very poorly with respect to number of nodes.

```
from tvm import relay
import mxnet as mx
from mxnet.gluon.model_zoo.vision import get_model
from tvm.relay import op as reg

def _register_external_op_helper(op_name, supported=True):
    @reg.register(op_name, "target.test")
    def _func_wrapper(attrs, args):
        return supported
    return _func_wrapper

_register_external_op_helper("nn.conv2d")
_register_external_op_helper("nn.dense")
_register_external_op_helper("nn.relu")
_register_external_op_helper("add")
_register_external_op_helper("multiply")
_register_external_op_helper("nn.bias_add")
_register_external_op_helper("nn.batch_flatten")
_register_external_op_helper("nn.max_pool2d")
_register_external_op_helper("nn.dropout")
_register_external_op_helper("nn.batch_norm")
_register_external_op_helper("nn.global_avg_pool2d")
_register_external_op_helper("clip")
_register_external_op_helper("concatenate")
_register_external_op_helper("nn.avg_pool2d")

def test_densenet():
    block = get_model('densenet121', pretrained=True)
    mod, params = relay.frontend.from_mxnet(block, shape={'data': (1, 3, 224, 224)}, dtype='float32')
    mod = relay.transform.AnnotateTarget("test")(mod)
    print('Running MergeCompilerRegions...')
    mod = relay.transform.MergeCompilerRegions()(mod)
    mod = relay.transform.PartitionGraph()(mod)
    print(mod)

test_densenet()
```

**Comment by mbaret** [comment]
Thanks again Trevor for testing :) That test completes for me (in less than a  second). If I had to take a guess, I'd say you might be hitting a recursion limit?

**Comment by mbaret** [comment]
Actually, are you using your patch for Tuples? I'm testing it without that patch.

**Comment by trevor-m** [comment]
> Actually, are you using your patch for Tuples? I'm testing it without that patch.

Yes, I am using that patch. Without the patch it works.

**Pull request merged**